feature/SecondDeliverable: Functional version 1.0.0

ponchOlvera opened the pull request:
+ Implemented Hilt Dependency Injection in severe modules
+ Added navigation graph and functionality as the navigation schema for the application, such as actions for transitions and safe arguments
+ Added detail screen, showing the details of a selected character, the comics and series as well.
+ Transformed two single retrofit requests into a unique single observer, transforming the data to match a single transformed object.
+ Added an onClick listener in the characters recycler view and enable the navigation to the details screen.
+ Added empty search fragment.
+ Added button navigation bar, and associate it with a navigation graph.

ponchOlvera committed:
feature/SecondDeliverable: V-0.0.1

+ Added Hilt Dependency injection.
+ Change classes to include the correct dependency pattern.
+ Removed unused imports.
+ Created Module classes to hold the creation of provider and bind methods for Hilt.
+ Code formatted.

ponchOlvera committed:
feature/SecondDeliverable: V-0.0.2

+ Implemented Hilt Dependency Injection in severe modules
+ Added navigation graph and functionality as the navigation schema for the application, such as actions for transitions and safe arguments
+ Added detail screen, showing the details of a selected character, the comics and series as well.
+ Transformed two single retrofit requests into a unique single observer, transforming the data to match a single transformed object.
+ Added an onClick listener in the characters recycler view and enable the navigation to the details screen.
+ Added empty search fragment.
+ Added button navigation bar, and associate it with a navigation graph.

pestrada commented on app/src/main/java/com/wizeline/heroes/ui/fragments/CharactersFragment.kt:
```diff
@@ -2,34 +2,42 @@ package com.wizeline.heroes.ui.fragments
+    @MarvelCharacters
+    @Inject
+    lateinit var marvelViewModelFactory: ViewModelProvider.Factory
+    var mMarvelViewModel: IMarvelViewModel? = null
```
This works. However, you should be able to use directly:
`val marvelViewModel: MarvelViewModel  by viewModels()`

Otherwise, you're just adding another level of indirection (the `IMarvelViewModel` interface) but it's not really needed. Also, by using `viewModels()` the view model will not be `null` at runtime.

ponchOlvera replied:
Done.

pestrada commented on app/src/main/java/com/wizeline/heroes/ui/fragments/CharactersFragment.kt:
```diff
@@ -62,8 +70,11 @@ class CharactersFragment : Fragment() {
         // Retrieving characters from retrofit
-        _marvelViewModel = ViewModelProvider(this).get(MarvelViewModel::class.java)
-        mMarvelViewModel.marvelViewState.observe(viewLifecycleOwner, {
+        val marvelViewModel: IMarvelViewModel by viewModels {
```
Based on Android docs: https://developer.android.com/training/dependency-injection/hilt-jetpack#viewmodels

the call to `by viewModels` should be at the class level (as a class property), not inside a function. This will ensure the viewModel is properly compiled and Dependencies generated at Compile Time, instead of Runtime.

ponchOlvera replied:
Done.

pestrada commented on app/src/main/java/com/wizeline/heroes/ui/fragments/SearchFragment.kt:
```diff
@@ -0,0 +1,63 @@
+        savedInstanceState: Bundle?
+    ): View {
+        _binding = FragmentSearchBinding.inflate(inflater, container, false)
+        //init()
```
let's remove commented code to ensure quality.

ponchOlvera replied:
Done.

pestrada commented on app/src/main/java/com/wizeline/heroes/viewmodels/IMarvelDetailsViewModel.kt:
```diff
@@ -0,0 +1,56 @@
+import javax.inject.Inject
+
+@HiltViewModel
+class MarvelDetailsViewModel @Inject constructor(val characterDetailsUseCase: MarvelCharacterDetailsUseCase) :
```
The file name should be the same as the class name -> `MarvelDetailsViewModel `
to avoid ambiguity in the code base.

ponchOlvera replied:
Done.

pestrada commented on app/src/main/java/com/wizeline/heroes/viewmodels/IMarvelDetailsViewModel.kt:
```diff
@@ -0,0 +1,56 @@
+    private var _marvelDetailsViewState = MutableLiveData(MarvelCharacterDetailsViewState())
+    val marvelDetailsViewState get() = _marvelDetailsViewState
+
+    private val compositeDisposable = CompositeDisposable()
```
Remember to call `compositeDisposable.clear()` on the ViewModels' `onCleared` method to ensure resources are cleaned up from memory when the ViewModel is destroyed.

ponchOlvera replied:
Done.

pestrada reviewed (commented):
Left a few comments but it's looking good 👍

ponchOlvera committed:
feature/SecondDeliverable: Functional version 1.0.1

+ Changed the instance method of the viewmodels in the fragments.
+ Removed manual implementation of the viewModel Factory
+ Renamed classes to match the filename.
+ Removed commented code.
+ Removed unused imports.

pestrada reviewed (approved):
Looks good 👍 Now you can remove the following files, since they're not needed:
- ViewModelFactoryModule
- IMarvelFragmentViewModel

Feel free to merge once you remove them.

ponchOlvera committed:
feature/SecondDeliverable: Functional version 1.0.2

+ Removed unused classes and DI modules.
+ Refactor viewmodels and not use interface implementation.
